Add support for handling pre-releases

mtreinish opened the pull request:
This commit adds support to the release automation for dealing with
pre-releases. A pre-release tag should trigger branch creation (if
enabled) but not trigger a metapackage bump PR. This commit adds the
packaging dependency to determine if a PEP440 comatible pre-release
is tagged and if so handle the tag as a pre-release. The pre-release
event will create a prerelease github release page with the changelog
from the appropriate previous full release and create a branch from the
tag if the repo is configured to branch on release.

Fixes #21

mtreinish committed:
Add support for handling pre-releases

This commit adds support to the release automation for dealing with
pre-releases. A pre-release tag should trigger branch creation (if
enabled) but not trigger a metapackage bump PR. This commit adds the
packaging dependency to determine if a PEP440 comatible pre-release
is tagged and if so handle the tag as a pre-release. The pre-release
event will create a prerelease github release page with the changelog
from the appropriate previous full release and create a branch from the
tag if the repo is configured to branch on release.

Fixes #21

mtreinish committed:
Merge branch 'master' into pre-release-support

jakelishman commented on qiskit_bot/release_process.py:
```diff
@@ -234,14 +235,16 @@ def _generate_changelog(repo, log_string, categories, show_missing=False):
-def _get_log_string(version_number_pieces):
-    version_number = '.'.join(version_number_pieces)
+def _get_log_string(version_number_pieces, version_string):
+    version_number = version_string
```
I'm not about to hold up the PR over it, but since the caller of this function has the parsed `packaging.version` object, we could just pass that as a singleton object into this helper function and also get rid of all the magic numbers in the indexing below.

mtreinish replied:
Yeah that's a good idea I'll do this, it would be a lot cleaner

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit-bot/pull/24/commits/255d475c38d60881eae87c2a9b813a30480c1fbc

jakelishman commented on tests/test_release_process.py:
```diff
@@ -853,16 +853,34 @@ def test_bump_minor_release_from_pending_patch_release_pr(self,
+        version_pieces = ['0', '25', '0']
+        self.assertEqual('0.25.0b1...0.24.0',
+                         release_process._get_log_string(version_pieces,
+                                                         "0.25.0b1"))
```
Do we definitely want `0.24.0` to be the "version before" `0.25.0rc2`, or should it be `0.25.0rc1` for release-note generation?

mtreinish replied:
I wasn't sure exactly how we wanted to handle this. This is just for the changelog in the release page like: https://github.com/Qiskit/qiskit-terra/releases/tag/0.20.0

For rc1 and the full release I'm pretty sure we want the full thing but rc2 I could see the value in making it strictly incremental over rc1. I don't really have a strong opinion as I never really understood why people like the changelog view over the full release notes (but some people do use it apparently)

jakelishman replied:
I think for a changelog, `rc2` should only be changes above `rc1`, probably.  The full release notes will always have everything, and I guess the type of person looking at the changelog for a second release candidate probably knows what was in the first release candidate already.

mtreinish replied:
I updated this in: https://github.com/Qiskit/qiskit-bot/pull/24/commits/255d475c38d60881eae87c2a9b813a30480c1fbc so it should show the changelog between prerelease tags.

mtreinish committed:
Use packaging Version object for log string generation

This commit simplifies the log string generation to use the parsed
packaging Version object to figure out the previous tag instead of
manually parsing the string.
